File: JavaScript/src/_00045_JumpGameII.js
/**
 * Given an array of integers `nums`, start at index 0. Each element `nums[i]`
 * represents the maximum length of a forward jump from index `i`. Return *the
 * minimum number of jumps to reach index* `n - 1`.
 * @param {number[]} nums array of integers that each represent the maximum
 * length of a forward jump from that index
 * @returns {number} the minimum number of jumps to reach the last index
 */
var jump = function (nums) {
    // the farthest index we can reach from positions in the current range
    let farthest = 0;
    // the farthest index we can reach using current number of jumps
    let currEnd = 0;
    // number of committed jumps
    let jumps = 0;

    // Stop jumping at second to last index because we don't need to jump
    // anymore once we reach the last index.
    for (let i = 0; i < nums.length - 1; i++) {
        // While scanning the current range, compute the best possible reach
        // for the next jump.
        farthest = Math.max(farthest, i + nums[i]);

        // If we've reached the end of the current jump's range, we must make
        // another jump.
        if (i === currEnd) {
            jumps++;
            // The new range extends to the farthest we've found.
            currEnd = farthest; 
        }
    }

    return jumps;
};

module.exports = { jump };

File: Python/_00045_JumpGameII.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        """Given an array of integers `nums`, start at position 0. Each element
        nums[i] represents the maximum length of a forward jump from index `i`.
        Return *the minimum number of jumps to reach the last index*.

        Args:
            nums (List[int]): array integers that represent the masimum length
            of a foward jump

        Returns:
            int: the minimum number of jumps needed to reach the end of the
            array
        """

        farthest = 0    # the farthest position that can be reached from an index
        jumps = 0       # the number of jumps made
        curr_end = 0     # the end of the possible range for the number of jumps

        for i in range(0, len(nums) - 1):
            farthest = max(farthest, i + nums[i])

            # If we reach the end of the range for the current number of jumps
            # we must jump again and establish a new range from the current
            # farthest reachable position.
            if i == curr_end:
                jumps += 1
                curr_end = farthest

        return jumps


import unittest


class TestSoluction(unittest.TestCase):
    def setUp(self) -> None:
        self.sol = Solution()

    def test_example1(self):
        self.assertEqual(self.sol.jump([2, 3, 1, 1, 4]), 2)

    def test_example2(self):
        self.assertEqual(self.sol.jump([2, 3, 0, 1, 4]), 2)

    def test_one_position_zero(self):
        self.assertEqual(self.sol.jump([0]), 0)

    def test_two_positions(self):
        self.assertEqual(self.sol.jump([1, 0]), 1)


if __name__ == "__main__":
    unittest.main()
